// @cardano-sdk/key-management/dist/cjs/cip8/coseUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCoseKeyCbor = exports.createCoseSign1Cbor = exports.createProtectedHeadersCbor = exports.COSE_HEADER_ALG = exports.COSE_KEY_X = exports.COSE_KEY_CRV = exports.COSE_KEY_ALG = exports.COSE_KEY_KID = exports.COSE_KEY_KTY = exports.CURVE_ED25519 = exports.KEY_TYPE_OKP = exports.ALGORITHM_EDDSA = void 0;
const core_1 = require("@cardano-sdk/core");
exports.ALGORITHM_EDDSA = -8;
exports.KEY_TYPE_OKP = 1;
exports.CURVE_ED25519 = 6;
exports.COSE_KEY_KTY = 1;
exports.COSE_KEY_KID = 2;
exports.COSE_KEY_ALG = 3;
exports.COSE_KEY_CRV = -1;
exports.COSE_KEY_X = -2;
exports.COSE_HEADER_ALG = 1;
const createProtectedHeadersCbor = (addressBytes) => {
    const writer = new core_1.Serialization.CborWriter();
    writer.writeStartMap(2);
    writer.writeInt(exports.COSE_HEADER_ALG);
    writer.writeInt(exports.ALGORITHM_EDDSA);
    writer.writeTextString('address');
    writer.writeByteString(addressBytes);
    return writer.encode();
};
exports.createProtectedHeadersCbor = createProtectedHeadersCbor;
const createCoseSign1Cbor = (protectedHeadersBytes, payloadBytes, signatureBytes, isHashed) => {
    const writer = new core_1.Serialization.CborWriter();
    writer.writeStartArray(4);
    writer.writeByteString(protectedHeadersBytes);
    writer.writeStartMap(1);
    writer.writeTextString('hashed');
    writer.writeBoolean(isHashed);
    writer.writeByteString(payloadBytes);
    writer.writeByteString(signatureBytes);
    return writer.encode();
};
exports.createCoseSign1Cbor = createCoseSign1Cbor;
const createCoseKeyCbor = (addressBytes, publicKeyHex) => {
    const writer = new core_1.Serialization.CborWriter();
    writer.writeStartMap(5);
    writer.writeInt(exports.COSE_KEY_KTY);
    writer.writeInt(exports.KEY_TYPE_OKP);
    writer.writeInt(exports.COSE_KEY_KID);
    writer.writeByteString(addressBytes);
    writer.writeInt(exports.COSE_KEY_ALG);
    writer.writeInt(exports.ALGORITHM_EDDSA);
    writer.writeInt(exports.COSE_KEY_CRV);
    writer.writeInt(exports.CURVE_ED25519);
    writer.writeInt(exports.COSE_KEY_X);
    writer.writeByteString(Buffer.from(publicKeyHex, 'hex'));
    return writer.encode();
};
exports.createCoseKeyCbor = createCoseKeyCbor;
